sqlglot/optimizer/simplify.py

```diff
@@ -430,6 +432,74 @@ def simplify_parens(expression):
     return expression
 
 
+CONSTANTS = (
+    exp.Literal,
+    exp.Boolean,
+    exp.Null,
+)
+
+
+def simplify_coalesce(expression):
+    # COALESCE(x) -> x
+    if (
+        isinstance(expression, exp.Coalesce)
+        and not expression.expressions
+        # COALESCE is also used as a Spark partitioning hint
+        and not isinstance(expression.parent, exp.Hint)
+    ):
+        return expression.this
+
+    if not isinstance(expression, COMPARISONS):
+        return expression
+
+    if isinstance(expression.left, exp.Coalesce):
+        coalesce = expression.left
+        other = expression.right
+    elif isinstance(expression.right, exp.Coalesce):
+        coalesce = expression.right
+        other = expression.left
+    else:
+        return expression
+
+    # COALESCE(x, y, z)
+    #                ^ grab that guy
+    coalesce_args = coalesce.expressions
+    if coalesce_args:
+        rightmost_arg = coalesce_args[-1]
```

Comment:
don't you want to iterate and fight the constant? not necessarily the last one,

coalesce(x, 1, 2) 

yes this is bad sql but you wanna ignore the 2 anyway

Reply:
This already simplifies that correctly, because we don't pop off the COALESCE.

```sql
COALESCE(x, 1, 2) = 2
```
Becomes
```sql
(
  NOT COALESCE(x, 1) IS NULL
  AND COALESCE(x, 1) = 2
)
OR (
  COALESCE(x, 1) IS NULL
  AND 2 = 2
)
```
Becomes
```sql
(
  NOT COALESCE(x, 1) IS NULL
  AND COALESCE(x, 1) = 2
)
OR (
  COALESCE(x, 1) IS NULL
)
```
Becomes
```sql
(
  NOT (
    (
      NOT COALESCE(x) IS NULL
      AND COALESCE(x) IS NULL
    ) OR (
      COALESCE(x) IS NULL
      AND 1 IS NULL
    )
  )
  AND (
    (
      NOT COALESCE(x) IS NULL
      AND COALESCE(x) = 2
    ) OR (
      COALESCE(x) IS NULL
      AND 1 = 2
    )
  )
)
OR (
  (
    NOT COALESCE(x) IS NULL
    AND COALESCE(x) IS NULL
  ) OR (
    COALESCE(x) IS NULL
    AND 1 IS NULL
  )
)
```
Becomes
```sql
NOT x IS NULL AND x = 2
```

Very roundabout haha
But kinda elegant...

Reply:
Its not so bad to find the first constant arg. So I'll just do that.